services/api-server/src/simcore_service_api_server/services/log_streaming.py

```diff
@@ -0,0 +1,135 @@
+
+    async def __aenter__(self):
+        await self.setup()
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.teardown()
+
+    async def _distribute_logs(self, data: bytes):
+        got = LoggerRabbitMessage.parse_raw(data)
+        item = JobLog(
+            job_id=got.project_id,
+            node_id=got.node_id,
+            log_level=got.log_level,
+            messages=got.messages,
+        )
+        callback = self._log_streamers.get(item.job_id)
+        if callback is None:
+            raise LogStreamerNotRegistered(
+                f"Could not forward log because a logstreamer associated with job_id={item.job_id} was not registered"
+            )
+        await callback(item)
+        return True
+
+    async def register(
+        self, job_id: JobID, callback: Callable[[JobLog], Awaitable[None]]
+    ):
+        if job_id in self._log_streamers:
+            raise LogStreamerRegistionConflict(
+                f"A stream was already connected to {job_id=}. Only a single stream can be connected at the time"
+            )
+        self._log_streamers[job_id] = callback
+        await self._rabbit_client.add_topics(
+            LoggerRabbitMessage.get_channel_name(), topics=[f"{job_id}.*"]
+        )
+
+    async def deregister(self, job_id: JobID):
+        if job_id not in self._log_streamers:
+            raise LogStreamerNotRegistered(f"No stream was connected to {job_id=}.")
+        await self._rabbit_client.remove_topics(
+            LoggerRabbitMessage.get_channel_name(), topics=[f"{job_id}.*"]
+        )
+        del self._log_streamers[job_id]
+
+
+class LogStreamer:
+    def __init__(
+        self,
+        user_id: UserID,
+        director2_api: DirectorV2Api,
+        job_id: JobID,
+        log_distributor: LogDistributor,
+    ):
+        self._user_id = user_id
+        self._director2_api = director2_api
+        self._queue: Queue[JobLog] = Queue()
+        self._job_id: JobID = job_id
+        self._log_distributor: LogDistributor = log_distributor
+        self._is_registered: bool = False
+
+    async def setup(self):
+        await self._log_distributor.register(self._job_id, self._queue.put)
+        self._is_registered = True
+
+    async def teardown(self):
+        await self._log_distributor.deregister(self._job_id)
+        self._is_registered = False
+
+    async def __aenter__(self):
+        await self.setup()
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.teardown()
+
+    async def _project_done(self) -> bool:
+        task = await self._director2_api.get_computation(self._job_id, self._user_id)
+        return not task.stopped is None
+
+    async def log_generator(self) -> AsyncIterable[str]:
+        if not self._is_registered:
+            raise LogStreamerNotRegistered(
+                f"LogStreamer for job_id={self._job_id} is not correctly registered"
+            )
+        while True:
```

Comment:
I think you could simplify a bit by using the return value of asyncio.sleep(_SLEEP_SECONDS_BEFORE_CHECK_JOB_STATUS, result=True)

Reply:
Hmm, I only what to do this when the queue is empty. If I understand your suggestion it would be to put the sleep in the outer loop. But, when the queue is not empty I want to stream those logs back to the client without waiting. So I think the sleep should be in the inner loop.